gloo/transport/uv/device.cc

```diff
@@ -0,0 +1,449 @@
+  hints.ai_socktype = SOCK_STREAM;
+  struct addrinfo* result;
+  auto rv = getaddrinfo(attr.hostname.data(), nullptr, &hints, &result);
+  GLOO_ENFORCE_NE(rv, -1);
+  struct addrinfo* rp;
+  for (rp = result; rp != nullptr; rp = rp->ai_next) {
+    auto fd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+    if (fd == -1) {
+      continue;
+    }
+
+    rv = bind(fd, rp->ai_addr, rp->ai_addrlen);
+    if (rv == -1) {
+      close(fd);
+      continue;
+    }
+
+    attr.ai_family = rp->ai_family;
+    attr.ai_socktype = rp->ai_socktype;
+    attr.ai_protocol = rp->ai_protocol;
+    memcpy(&attr.ai_addr, rp->ai_addr, rp->ai_addrlen);
+    attr.ai_addrlen = rp->ai_addrlen;
+    close(fd);
+    break;
+  }
+
+  // Check that we found an address we were able to bind to
+  GLOO_ENFORCE(rp != nullptr, "Unable to find address for: ", attr.hostname);
+  freeaddrinfo(result);
+  return;
+}
+
+} // namespace
+
+std::shared_ptr<transport::Device> CreateDevice(struct attr attr) {
+  if (!attr.iface.empty()) {
+    // Initialize attributes using network interface name
+    if (!lookupAddrForIface(&attr)) {
+      GLOO_ENFORCE(false, "Unable to find address for: ", attr.iface);
+    }
+  } else {
+    // Initialize attributes using hostname/IP address
+    // If not already specified, use this machine's hostname
+    if (attr.hostname.size() == 0) {
+      std::array<char, HOST_NAME_MAX> hostname;
+      auto rv = gethostname(hostname.data(), hostname.size());
+      GLOO_ENFORCE_EQ(rv, 0);
+      attr.hostname = hostname.data();
+    }
+    lookupAddrForHostname(attr);
+  }
+
+  return std::make_shared<Device>(attr);
+}
+
+Device::Device(const struct attr& attr) : attr_(attr) {
+  int rv;
+
+  loop_ = libuv::Loop::create();
+
+  // Use async handle to trigger the event loop to
+  // run deferred functions on its thread.
+  async_ = loop_->resource<libuv::Async>();
+  async_->on<libuv::AsyncEvent>(
+      [this](const libuv::AsyncEvent&, const libuv::Async&) {
+        this->asyncCallback();
+      });
+
+  // Initialize server handle and wait for incoming connections.
+  listener_ = loop_->resource<libuv::TCP>();
+  listener_->on<libuv::ErrorEvent>(
+      [this](const libuv::ErrorEvent& event, const libuv::TCP&) {
+        // Nothing we can do about errors on the listener socket...
+        GLOO_ENFORCE(!event, "Error on listener socket: ", event.what());
+      });
+  listener_->on<libuv::ListenEvent>(
+      [this](const libuv::ListenEvent& event, const libuv::TCP&) {
+        listenCallback();
+      });
+
+  // Bind socket and start listening for new connections.
+  listener_->bind((const struct sockaddr*)&attr_.ai_addr);
+  listener_->listen();
+  addr_ = Address(listener_->sockname());
+
+  // Run uv_run on private thread.
+  thread_.reset(new std::thread([this] { loop_->run(); }));
+}
+
+Device::~Device() {
+  // Close handles associated with this device.
+  defer([this] {
+    listener_->close();
+    async_->close();
```

Comment:
The `async_` handle is just something to poke the event loop. It is used to wake up the event loop when a new lambda has been added to the deque of deferred lambda's (to be executed on the event loop thread).

They can be used by pointer reference (through `this` here), because in the statement below the destructor waits for the event loop thread to finish (via `thread::join`). This is necessary because the thread is a `std::thread` member variable on the device instance and therefore must finish before the object can be destructed.

You're right that if we wouldn't wait for the thread to join, we'd have to capture a whole lot more in the lambda that we defer, as well as somehow wait for the event loop to pick it up, as well as unregister all the event handlers that use the device instance.